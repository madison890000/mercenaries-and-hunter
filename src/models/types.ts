import StringWithID from './StringWithID';

namespace DataModel {
    export enum Degree {
        Bachelor = 'BACHELOR',
        Master = 'MASTER',
        PhD = 'PHD'
    }

    export type Gender = 'MALE' | 'FEMALE';

    export type IAchievement = {
        text: string;
        categories: string[];
    };
    export type Achievement = {
        id: string;
        text: string;
        categories: StringWithID[];
    };
    export type SolutionsOfHowToImplement = string;

    export enum JobPositionLevel {
        Staff = 'STAFF',
        Senior = 'SENIOR',
        Middle = '',
        Junior = 'JUNIOR'
    }

    export type JobPosition = 'Software Engineer' | 'Front-End Engineer' | 'Back-End Engineer' | 'Full-Stack Engineer';

    export type Job = {
        position: JobPosition;
        level: JobPositionLevel;
    };
    export enum SkillLevel {
        understand = 'UNDERSTAND',
        familiar = 'FAMILIAR',
        proficient = 'PROFICIENT'
    }

    export enum Importance {
        Essential,
        Advanced
    }

    export enum CompanyType {
        Enterprise,
        Startup
    }
}
export default DataModel;
