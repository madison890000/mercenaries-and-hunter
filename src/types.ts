export enum SendType {
    SEND = 0,
    START_INTERVIEWING = 1,
    MIDDLE_INTERVIEWING = 2,
    FINAL_INTERVIEWING = 3,
    OFFER = 4,
    REJECTED = 5,
}

export interface IAppliedInExtension {
    site: string;
    time: string;
    id: string;
    title: string;
    originUrl?: string
}

export interface IAppliedInLocal extends IAppliedInExtension {
    status: SendType;
    like: number;
    needSync: boolean;
    serverId?: number;
    updateTime?: number;
}

export interface IAppliedInServer extends Omit<IAppliedInLocal, 'id'> {
    status: SendType;
    id: number;
    like: number;
    needSync: boolean;
    serverId: number;
    updateTime: number;
}

export type NeedSyncAppliedItem = Omit<IAppliedInLocal, 'id'> & { id?: number }