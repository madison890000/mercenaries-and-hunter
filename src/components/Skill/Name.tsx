import React, { PropsWithChildren } from 'react';
import styles from './Name.module.scss';
import DataModel from '../../Model/types';

const Name: React.FC<PropsWithChildren<{ level: DataModel.Importance }>> = ({ level, children }) => {
    switch (level) {
        case DataModel.Importance.important:
            return <div className={styles.important}>{children}</div>;
        case DataModel.Importance.normal:
            return <div className={styles.normal}>{children}</div>;
        case DataModel.Importance.less:
            return <div className={styles.less}>{children}</div>;
        default:
            return null;
    }
};

export default Name;
