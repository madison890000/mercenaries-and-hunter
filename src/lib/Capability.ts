export default class Capability {
    public coding: number;
    public leaning: number;
    public algorithm: number;
    public communicate: number;
    public engineer: number;
    public passion: number;

    constructor({
                    coding,
                    leaning,
                    algorithm,
                    communicate,
                    engineer,
                    passion
                }: any) {
        this.coding = coding;
        this.leaning = leaning;
        this.algorithm = algorithm;
        this.communicate = communicate;
        this.engineer = engineer;
        this.passion = passion;
    }

}
