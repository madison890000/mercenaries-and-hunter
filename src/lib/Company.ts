interface ICompany {
    name: string;
    location?: string;
}

export default class Company {
    public name: string;
    public location?: string;

    constructor({ name, location }: ICompany) {
        this.name = name;
        this.location = location;
    }

}
