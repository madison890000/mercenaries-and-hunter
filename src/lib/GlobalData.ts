import { getItem, saveItem } from '../utils';

class GlobalStore {
  private keys: string[];
  private data: any;

  constructor() {
    this.keys = [];
    this.data = {};
    this.recover();
  }

  private static DATA_KEYS = 'global-data-key';

  save(key: string, value: any) {
    if (!this.keys?.find(k => k === key)) {
      this.keys.push(key);
      saveItem(GlobalStore.DATA_KEYS, JSON.stringify(this.keys));
    }
    this.data[key] = value;
    saveItem(key, JSON.stringify(value));
  }

  get(key: string) {
    return this.data[key];
  }

  async syncGetWithNoCache(key: string) {
    const values = (await getItem(key)) as any;
    this.data[key] = JSON.parse(values);
    return this.data[key];
  }

  recover = async () => {
    const data = (await getItem(GlobalStore.DATA_KEYS)) as any[];
    try {
      if (typeof data === 'string') {
        this.keys = JSON.parse(data);
      }
    } catch (e) {
      console.log(e);
    }
    for (const key of this.keys) {
      const values = (await getItem(key)) as any;
      this.data[key] = JSON.parse(values);
    }
  };
}

const globalStore = new GlobalStore();
export default globalStore;
