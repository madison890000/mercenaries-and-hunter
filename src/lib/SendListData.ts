import {syncSends} from "../services/mh";
import {IS_DEV} from "../constants/environment";
import {getItem, saveItem} from "../utils";

class SendList {
    private localData: any[];
    private serviceData: any[];

    constructor() {
        this.localData = [];
        this.serviceData = [];
        this.recover();
    }

    private static DATA_KEY = 'send-list-data';

    updateLocalDataByIDAndField(id: string, field: any, saveNow: boolean) {
        const index = this.localData?.findIndex((e) => e?.id === id);
        const localData = [...this.localData];
        localData[index] = {
            ...this.localData[index],
            ...field,
        };
        this.localData = localData;
        saveNow && this.save();
    }

    getNewOne = (a: any, b: any, defaultOne: any) => {
        if (a?.updateTime && b?.updateTime) {
            return a?.updateTime > b?.updateTime ? a : b;
        } else if (!a?.updateTime && b?.updateTime) {
            return b
        } else if (a?.updateTime && !b?.updateTime) {
            return a
        } else if (!a?.updateTime && !b?.updateTime) {
            return defaultOne
        }
    }

    diffLastRestoreAndServerData() {
        const last = this.localData;
        const server = this.serviceData;
        last?.forEach(l => {
            const find = server?.some(s => {
                if (l.site === s.site) {
                    if (l.like !== s?.like || l.status !== s?.status) {
                        const saveData = this.getNewOne(l, s, l);
                        this.updateLocalDataByIDAndField(l?.id, {
                            serverId: s?.id,
                            like: saveData?.like,
                            status: saveData?.status,
                            updateTime: saveData?.updateTime,
                            needSync: true,
                        }, false)
                    } else {
                        this.updateLocalDataByIDAndField(l?.id, {
                            serverId: s?.id,
                            needSync: false,
                        }, false)
                    }
                    return true
                } else {
                    return false
                }
            });
            if (!find) {
                this.updateLocalDataByIDAndField(l?.id, {
                    needSync: true,
                }, false)
            }
        });
    }

    setServerData(d: any[]) {
        this.serviceData = d;
        this.diffLastRestoreAndServerData();
        this.save();
    }


    syncDataToServer = async () => {
        const needSyncData = this.localData?.filter(e => e?.needSync);
        if (needSyncData?.length > 0) {
            await syncSends(needSyncData?.map(e => ({
                ...e,
                id: e?.serverId,
            })));
            needSyncData?.forEach(e => {
                this.updateLocalDataByIDAndField(e?.id, {
                    needSync: false
                }, false)
            })
            this.save();
        }
    }

    getListByIds(ids: string[]) {
        if (IS_DEV) {
            return this.localData
        }
        return this.localData?.filter(d => ids?.includes(d?.id))
    }

    addMore(data: any[]) {
        let changed = false;
        data?.forEach((d) => {
            if (!this.localData.find(a => a?.id === d?.id)) {
                const newData = {
                    ...d,
                    status: 0,
                    like: 3,
                    needSync: true,
                };
                this.localData.push(newData);
                changed = true;
            }
        });
        changed && this.save();
    }

    updateStatusById(id: string, status: any) {
        this.updateLocalDataByIDAndField(id, {
            status,
            updateTime: new Date().getTime(),
            needSync: true,
        }, true)
    }

    updateLikeById(id: string, like: any) {
        this.updateLocalDataByIDAndField(id, {
            like,
            updateTime: new Date().getTime(),
            needSync: true,
        }, true)
    }

    save() {
        saveItem(SendList.DATA_KEY, JSON.stringify(this.localData))
    }


    recover = async () => {
        const data = await getItem(SendList.DATA_KEY) as any[];
        try {
            if (typeof data === "string") {
                this.localData = JSON.parse(data);
            } else {
                this.localData = data || [];
            }
        } catch (e) {
            console.log(e);
        }
    }

}

const sendList = new SendList();
export default sendList