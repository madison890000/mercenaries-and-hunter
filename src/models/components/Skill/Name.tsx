import React, {PropsWithChildren} from 'react';
import styles from './Name.module.scss';
import {Importance} from "../../types";

const Name: React.FC<PropsWithChildren<{ level: Importance }>> = ({level, children}) => {
    switch (level) {
        case Importance.Essential:
            return <div className={styles.important}>{children}</div>;
        case Importance.Advanced:
            return <div className={styles.normal}>{children}</div>;
        default:
            return null;
    }
};

export default Name;
