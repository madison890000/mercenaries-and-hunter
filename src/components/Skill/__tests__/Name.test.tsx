import React from 'react';
import Name from '../Name';
import renderer from 'react-test-renderer';
import { Importance } from '../../../lib/types';

describe('renders Name ', () => {

    test('Name --->level = important', () => {
        const nameImportant = renderer.create(
            <Name
                level={Importance.important}
            >understand normal Skill</Name>
        )
            .toJSON();
        expect(nameImportant)
            .toMatchSnapshot();
    });

    test('Name --->level = normal', () => {
        const nameNormal = renderer.create(
            <Name
                level={Importance.normal}
            >understand normal Skill</Name>
        )
            .toJSON();
        expect(nameNormal)
            .toMatchSnapshot();
    });

    test('Name --->level = less', () => {
        const nameLess = renderer.create(
            <Name
                level={Importance.less}
            >understand normal Skill</Name>
        )
            .toJSON();
        expect(nameLess)
            .toMatchSnapshot();
    });

})
