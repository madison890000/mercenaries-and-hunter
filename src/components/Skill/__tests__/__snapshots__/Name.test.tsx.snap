// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Name  Name --->level = important 1`] = `null`;

exports[`renders Name  Name --->level = less 1`] = `null`;

exports[`renders Name  Name --->level = normal 1`] = `null`;
